internal/app/server/impl.go

```diff
@@ -0,0 +1,324 @@
+	responseDefinition := req.Msg.ResponseDefinition
+	if responseDefinition != nil {
+		addHeaders(responseDefinition.ResponseHeaders, stream.ResponseHeader())
+		addHeaders(responseDefinition.ResponseTrailers, stream.ResponseTrailer())
+	}
+
+	// Convert the request to an Any so that it can be recorded in the payload
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return err
+	}
+	requestInfo := createRequestInfo(req.Header(), []*anypb.Any{msgAsAny})
+	payload := &v1alpha1.ConformancePayload{
+		RequestInfo: requestInfo,
+	}
+
+	for _, data := range responseDefinition.ResponseData {
+		payload.Data = data
+
+		resp := &v1alpha1.ServerStreamResponse{
+			Payload: payload,
+		}
+
+		time.Sleep((time.Duration(responseDefinition.ResponseDelayMs) * time.Millisecond))
+
+		if err := stream.Send(resp); err != nil {
+			return connect.NewError(connect.CodeInternal, fmt.Errorf("error sending on stream: %w", err))
+		}
+		// Only echo back the request info in the first response
+		payload.RequestInfo = nil
+	}
+	if responseDefinition.Error != nil {
+		return createError(responseDefinition.Error)
+	}
+	return nil
+}
+
+func (s *conformanceServer) BidiStream(
+	ctx context.Context,
+	stream *connect.BidiStream[v1alpha1.BidiStreamRequest, v1alpha1.BidiStreamResponse],
+) error {
+	var responseDefinition *v1alpha1.StreamResponseDefinition
+	fullDuplex := false
+	firstRecv := true
+	respNum := 0
+	var reqs []*anypb.Any
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		req, err := stream.Receive()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Reads are done, break the receive loop and send any remaining responses
+				break
+			}
+			return fmt.Errorf("receive request: %w", err)
+		}
+
+		// Record all requests received
+		msgAsAny, err := asAny(req)
+		if err != nil {
+			return err
+		}
+		reqs = append(reqs, msgAsAny)
+
+		// If this is the first message in the stream, save off the total responses we need to send
+		// plus whether this should be full or half duplex
+		if firstRecv {
+			responseDefinition = req.ResponseDefinition
+			fullDuplex = req.FullDuplex
+			firstRecv = false
+		}
+
+		// If fullDuplex, then send one of the desired responses each time we get a message on the stream
+		if fullDuplex {
+			if respNum >= len(responseDefinition.ResponseData) {
+				return connect.NewError(
+					connect.CodeAborted,
+					errors.New("received more requests than desired responses on a full duplex stream"),
+				)
+			}
+			requestInfo := createRequestInfo(stream.RequestHeader(), reqs)
+			resp := &v1alpha1.BidiStreamResponse{
+				Payload: &v1alpha1.ConformancePayload{
+					RequestInfo: requestInfo,
+					Data:        responseDefinition.ResponseData[respNum],
+				},
+			}
+			time.Sleep((time.Duration(responseDefinition.ResponseDelayMs) * time.Millisecond))
+
+			if err := stream.Send(resp); err != nil {
+				return connect.NewError(connect.CodeInternal, fmt.Errorf("error sending on stream: %w", err))
+			}
```

Comment:
This could probably be pulled out into its own function also. Lmk if you think it is worth it.